graylog2-server/src/main/java/org/graylog2/database/PaginatedDbService.java

```diff
@@ -0,0 +1,234 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Graylog.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.graylog2.database;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Streams;
+import org.bson.types.ObjectId;
+import org.graylog2.bindings.providers.MongoJackObjectMapperProvider;
+import org.mongojack.DBCursor;
+import org.mongojack.DBQuery;
+import org.mongojack.DBSort;
+import org.mongojack.JacksonDBCollection;
+import org.mongojack.WriteResult;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * This class is a helper to implement a basic Mongojack-based database service that allows CRUD operations on a
+ * single DTO type and offers paginated access.
+ * <p>
+ *     It makes only a few assumptions, which are common to many Graylog entities:
+ *     <ul>
+ *         <li>The DTO class has a name which is unique</li>
+ *     </ul>
+ * </p>
+ * <p>
+ *     Subclasses can add more sophisticated query methods by access the protected "db" property.<br/>
+ *     Indices can be added in the constructor.
+ * </p>
+ * @param <DTO>
+ */
+public abstract class PaginatedDbService<DTO> {
+    protected final JacksonDBCollection<DTO, ObjectId> db;
+
+    protected PaginatedDbService(MongoConnection mongoConnection,
+                                 MongoJackObjectMapperProvider mapper,
+                                 Class<DTO> dtoClass,
+                                 String collectionName) {
+        this.db = JacksonDBCollection.wrap(mongoConnection.getDatabase().getCollection(collectionName),
+                dtoClass,
+                ObjectId.class,
+                mapper.get());
+    }
+
+    /**
+     * Get the {@link DTO} for the given ID.
+     *
+     * @param id the ID of the object
+     * @return an Optional containing the found object or an empty Optional if no object can be found for the given ID
+     */
+    public Optional<DTO> get(String id) {
+        return Optional.ofNullable(db.findOneById(new ObjectId(id)));
+    }
+
+    /**
+     * Stores the given {@link DTO} in the database.
+     *
+     * @param dto the {@link DTO} to save
+     * @return the newly saved {@link DTO}
+     */
+    public DTO save(DTO dto) {
+        final WriteResult<DTO, ObjectId> save = db.save(dto);
+        return save.getSavedObject();
+    }
+
+    /**
+     * Deletes the {@link DTO} for the given ID from the database.
+     *
+     * @param id ID of the {@link DTO} to delete
+     */
+    public void delete(String id) {
+        db.removeById(new ObjectId(id));
```

Comment:
Please return the `WriteResult` in this method so that callers can check it if required.

Reply:
Do you think we should return the complete `WriteResult<DTO, ObjectId` or would it be enough to return the value of `WriteResult.getN()`? This is what we do in other parts of the server and it would avoid leaking the `WriteResult` object to the caller.

Reply:
Sorry, somehow missed the second part of your sentence. I will return the number of deleted documents. :+1:

Reply:
✅